src/rpc/server.zig

```diff
@@ -0,0 +1,547 @@
+        std.mem.Allocator.Error ||
+        std.http.Server.ReceiveHeadError ||
+        AcceptConnectionError;
+
+    pub fn acceptAndServeConnection(server: *Server) AcceptAndServeConnectionError!void {
+        const conn = (try acceptConnection(&server.tcp, server.logger)).?;
+        errdefer conn.stream.close();
+
+        server.wait_group.start();
+        errdefer server.wait_group.finish();
+
+        const hct_buffer = try server.allocator.alignedAlloc(
+            u8,
+            @alignOf(HandleConnectionTask),
+            HandleConnectionTask.initBufferSize(server.read_buffer_size),
+        );
+        errdefer server.allocator.free(hct_buffer);
+
+        const new_hct = try HandleConnectionTask.initBufferAndReceiveHead(
+            hct_buffer,
+            server,
+            conn,
+        );
+        errdefer new_hct.http_server.connection.stream.close();
+
+        server.thread_pool.schedule(ThreadPool.Batch.from(&new_hct.task));
+    }
+};
+
+const HandleConnectionTask = struct {
+    task: ThreadPool.Task,
+    server: *Server,
+    http_server: std.http.Server,
+    request: std.http.Server.Request,
+
+    fn initBufferSize(read_buffer_size: usize) usize {
+        return @sizeOf(HandleConnectionTask) + read_buffer_size;
+    }
+
+    fn getFullBuffer(
+        read_buffer_size: usize,
+        hct: *HandleConnectionTask,
+    ) []align(@alignOf(HandleConnectionTask)) u8 {
+        const ptr: [*]align(@alignOf(HandleConnectionTask)) u8 = @ptrCast(hct);
+        return ptr[0..initBufferSize(read_buffer_size)];
+    }
+
+    fn getReadBuffer(
+        read_buffer_size: usize,
+        hct: *HandleConnectionTask,
+    ) []u8 {
+        return getFullBuffer(read_buffer_size, hct)[@sizeOf(HandleConnectionTask)..];
+    }
+
+    fn initBufferAndReceiveHead(
+        /// Must be `.len == initBufferSize(server.read_buffer_size)`.
+        buffer: []align(@alignOf(HandleConnectionTask)) u8,
+        server: *Server,
+        conn: std.net.Server.Connection,
+    ) std.http.Server.ReceiveHeadError!*HandleConnectionTask {
+        std.debug.assert(buffer.len == initBufferSize(server.read_buffer_size));
+
+        const hct: *HandleConnectionTask = std.mem.bytesAsValue(
+            HandleConnectionTask,
+            buffer[0..@sizeOf(HandleConnectionTask)],
+        );
+        hct.* = .{
+            .task = .{ .callback = callback },
+            .server = server,
+            .http_server = std.http.Server.init(conn, getReadBuffer(server.read_buffer_size, hct)),
+            .request = try hct.http_server.receiveHead(),
+        };
+
+        return hct;
+    }
+
+    fn callback(task: *ThreadPool.Task) void {
+        const hct: *HandleConnectionTask = @fieldParentPtr("task", task);
+        const server = hct.server;
+        const logger = server.logger;
+
+        const full_buffer = getFullBuffer(hct.server.read_buffer_size, hct);
+        defer hct.server.allocator.free(full_buffer);
```

Comment:
would be good to add a comment explaining how this free's the underlying task's memory

Reply:
In light of ce0c47543da8b6befabc76e4892e50154b0904ec, I think this is now clear just through the code itself, lmk wyt.